src/main/java/uk/gov/hmcts/reform/iacaseapi/domain/handlers/presubmit/DeriveHearingCentreHandler.java

```diff
@@ -131,18 +151,41 @@ private void trySetHearingCentreFromPostcode(
             String staffLocationName = StaffLocation.getLocation(hearingCentre).getName();
             asylumCase.write(STAFF_LOCATION, staffLocationName);
             asylumCase.write(CASE_MANAGEMENT_LOCATION,
-                caseManagementLocationService.getCaseManagementLocation(staffLocationName));
+                    caseManagementLocationService.getCaseManagementLocation(staffLocationName));
         } else {
             asylumCase.write(HEARING_CENTRE, hearingCentreFinder.getDefaultHearingCentre());
             asylumCase
-                .write(APPLICATION_CHANGE_DESIGNATED_HEARING_CENTRE, hearingCentreFinder.getDefaultHearingCentre());
+                    .write(APPLICATION_CHANGE_DESIGNATED_HEARING_CENTRE, hearingCentreFinder.getDefaultHearingCentre());
 
             String staffLocationName =
-                StaffLocation.getLocation(hearingCentreFinder.getDefaultHearingCentre()).getName();
+                    StaffLocation.getLocation(hearingCentreFinder.getDefaultHearingCentre()).getName();
             asylumCase.write(STAFF_LOCATION, staffLocationName);
             asylumCase.write(CASE_MANAGEMENT_LOCATION,
-                caseManagementLocationService.getCaseManagementLocation(staffLocationName));
+                    caseManagementLocationService.getCaseManagementLocation(staffLocationName));
         }
     }
 
+    public void setHearingCentreFromDetentionFacilityName(AsylumCase asylumCase) {
+        final String prisonName = asylumCase.read(PRISON_NAME, String.class).orElse("");
+
+        final String ircName = asylumCase.read(IRC_NAME, String.class).orElse("");
+
+        if (prisonName.isEmpty() && ircName.isEmpty()) {
+            throw new RequiredFieldMissingException("Prison name and IRC name missing");
+
+        } else {
+
+            String detentionFacilityName = !prisonName.isEmpty() ? prisonName : ircName;
+
+            HearingCentre hearingCentre = hearingCentreFinder.findByDetentionFacility(detentionFacilityName);
+            asylumCase.write(HEARING_CENTRE, hearingCentre);
+            asylumCase.write(APPLICATION_CHANGE_DESIGNATED_HEARING_CENTRE, hearingCentre);
+
+            String staffLocationName = StaffLocation.getLocation(hearingCentre).getName();
+            asylumCase.write(STAFF_LOCATION, staffLocationName);
```

Comment:
Do all HC have staff location or do we need to check?

Reply:
Checked and confirm that each hearingCentre has a StaffLocation mapping. Can also see this working correctly in CaseData